File: 프로그래머스/2/42626. 더 맵게/더 맵게.java
import java.util.PriorityQueue;
class Solution {
    public int solution(int[] scoville, int K) {
        PriorityQueue<Integer> q = new PriorityQueue<>();
        int answer = 0;
        for(int s : scoville){
            q.offer(s);
        }
     do {   
         if (q.peek() >= K){
                break;
        }
        int least = q.poll();
        int secondLeast = q.poll();
        int newSco = least + secondLeast*2;

        q.offer(newSco);

        answer ++;
        }
        while(q.size() > 1);
        
        if (q.peek() < K){
            return -1;
        }
        return answer;
    }
}

File: 프로그래머스/2/42626. 더 맵게/더 맵게.py
import heapq

class Scoville:
    def __init__(self, scoville, k):
        heapq.heapify(scoville)
        self.sco = scoville
        self.k = k
        self.count = 0
    def doNotMix(self):
        if self.sco[0] >= self.k:
            return 1
        else:
            return 0       
    def haveToMix(self):
        while self.doNotMix() == 0:
            if len(self.sco) < 2:
                return -1
        
            self.count += 1
            
            least = heapq.heappop(self.sco)
            
            secondLeast = heapq.heappop(self.sco)
            
            new = least + secondLeast * 2
            
            heapq.heappush(self.sco, new)
            
        return  self.count

def solution(scoville, k):
    sco = Scoville(scoville, k)
    answer = sco.haveToMix()

    return answer
